app/src/main/java/com/epfl/drawyourpath/mainpage/fragments/ChatFragment.kt

```diff
@@ -1,43 +1,251 @@
+                                    val welcomeMessageContent =
+                                        MessageContent.Text(getString(R.string.chat_welcome_message)) // Replace with your MessageContent creation
+                                    val welcomeMessage = Message(
+                                        0L,
+                                        userId,
+                                        welcomeMessageContent,
+                                        System.currentTimeMillis(),
+                                    )
+                                    val newChatPreview =
+                                        ChatPreview(conversationId, chatName, welcomeMessage)
+
+                                    // Add the new ChatPreview to the chatList and notify the adapter
+                                    requireActivity().runOnUiThread {
+                                        chatList.add(newChatPreview)
+                                        chatAdapter.notifyDataSetChanged()
+                                    }
+                                }
+                        }
+                    }
+                    true
+                }
+                popupMenu.show()
+            }.exceptionally { exception ->
+                Log.e(TAG, "Error while getting UserAccount: ", exception)
+                null
+            }
+        }
+    }
+
+    /**
+     * This function fetches the list of ChatPreview objects associated with a user from the database.
+     *
+     * @param database The Database object used to fetch user data and chat previews.
+     * @param userId The ID of the user whose chat previews are to be fetched.
+     * @return A CompletableFuture that contains a list of ChatPreview objects when completed.
+     */
+    fun getUserChatPreviews(
+        database: Database,
+        userId: String,
+    ): CompletableFuture<List<ChatPreview>> {
+        // Return a CompletableFuture that fetches the user data from the database
+        return database.getUserData(userId)
+            .thenCompose { userData ->
+                // Get the list of conversation IDs associated with the user
+                // If the user's chatList is null, use an empty list
+                val chatList = userData.chatList ?: emptyList()
+
+                // Create a list of CompletableFuture objects, each fetching a chat preview for a conversation ID
+                val chatPreviewFutures = chatList.map { conversationId ->
+                    database.getChatPreview(conversationId)
+                }
+
+                // Create a new CompletableFuture that completes when all chat preview futures complete
+                CompletableFuture.allOf(*chatPreviewFutures.toTypedArray())
+                    .thenApply {
+                        // Convert the list of futures to a list of ChatPreview objects
+                        chatPreviewFutures.map { it.join() }
+                    }

```

Comment:
It would be better not to use join(). It could be done by doing
```kotlin
.thenCompose {
     CompletableFuture.allOf(*chatPreviewFutures.toTypedArray())
}
```

Reply:
Im not sure what you mean. The above doesn't work as it is required that the futures are completed.

Reply:
Since the chatPrevies are in a thenApply functions, thy should be completed I think.

Reply:
Sorry for the misunderstanding, at the end of the thenCompose, you can apply a thenApply to access the list of previews:

```kotlin
.thenCompose {
     CompletableFuture.allOf(*chatPreviewFutures.toTypedArray())
}.thenApply{listMilestones ->
.....
}
```